tpm2/crypto_test.go

```diff
@@ -368,3 +373,365 @@ func TestPub(t *testing.T) {
+					FixedParent:          false,
+					SensitiveDataOrigin:  true,
+					UserWithAuth:         true,
+					AdminWithPolicy:      true,
+					NoDA:                 true,
+					EncryptedDuplication: false,
+					Restricted:           false,
+					Decrypt:              false,
+					SignEncrypt:          true,
+				},
+				AuthPolicy: duplicationPolicyHash(),
+				Parameters: NewTPMUPublicParms(
+					TPMAlgRSA,
+					&TPMSRSAParms{
+						KeyBits: 2048,
+						Scheme: TPMTRSAScheme{
+							Scheme: TPMAlgRSASSA,
+							Details: NewTPMUAsymScheme(
+								TPMAlgRSASSA,
+								&TPMSSigSchemeRSASSA{
+									HashAlg: TPMAlgSHA256,
+								},
+							),
+						},
+					},
+				),
+			}),
+		},
+		{
+			name: "ecdsa_p256",
+			alg:  TPMAlgECDSA,
+			pub: New2B(TPMTPublic{
+				Type:    TPMAlgECC,
+				NameAlg: TPMAlgSHA256,
+				ObjectAttributes: TPMAObject{
+					FixedTPM:             false,
+					STClear:              false,
+					FixedParent:          false,
+					SensitiveDataOrigin:  true,
+					UserWithAuth:         true,
+					AdminWithPolicy:      true,
+					NoDA:                 true,
+					EncryptedDuplication: false,
+					Restricted:           false,
+					Decrypt:              false,
+					SignEncrypt:          true,
+				},
+				AuthPolicy: duplicationPolicyHash(),
+				Parameters: NewTPMUPublicParms(
+					TPMAlgECC,
+					&TPMSECCParms{
+						CurveID: TPMECCNistP256,
+						Scheme: TPMTECCScheme{
+							Scheme: TPMAlgECDSA,
+							Details: NewTPMUAsymScheme(
+								TPMAlgECDSA,
+								&TPMSSigSchemeECDSA{
+									HashAlg: TPMAlgSHA256,
+								},
+							),
+						},
+					},
+				),
+			}),
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			// Create an exportable signing key and then export it.
+			create, err := Create{
+				ParentHandle: NamedHandle{
+					Handle: srkCreate.ObjectHandle,
+					Name:   srkCreate.Name,
+				},
+				InPublic: tc.pub,
+			}.Execute(tpm)
+			if err != nil {
+				t.Fatalf("could not create key: %v", err)
+			}
+
+			load, err := Load{
+				ParentHandle: NamedHandle{
+					Handle: srkCreate.ObjectHandle,
+					Name:   srkCreate.Name,
+				},
+				InPrivate: create.OutPrivate,
+				InPublic:  create.OutPublic,
+			}.Execute(tpm)
+			if err != nil {
+				t.Fatalf("could not load key: %v", err)
+			}
+			defer FlushContext{
+				FlushHandle: load.ObjectHandle,
+			}.Execute(tpm)
+
+			pub, priv := duplicateAndImport(t, tpm, load.ObjectHandle)
+			checkSignVerify(t, tpm, NamedHandle{
+				Handle: load.ObjectHandle,
+				Name:   load.Name,
+			}, tc.alg, pub, priv)
+		})
+	}
+}
+
+// Uses the `Pub` and `Priv` helpers to import the public and private key.
+// Note that in practice, one could just use `crypto.PrivateKey.Public()` to
+// derive the public key from the private key. This helper uses the `Pub`
+// routine to validate that function specifically while we test the rest.
+func duplicateAndImport(t *testing.T, tpm transport.TPM, h TPMHandle) (crypto.PublicKey, crypto.PrivateKey) {
```

Comment:
Seems like this function duplicates and exports the key, but it never "import" the key back to the TPM, maybe need to updat e the function name?

Reply:
Oh yeah, I'll clarify. It "imports" it to software :)